### PR TITLE
CASSANDRA-20667: Accord: BEGIN TRANSACTIONs IF condition logic does not properly support meaningless emptiness and null values

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: src/java/org/apache/cassandra/cql3/conditions/ColumnCondition.java [hunk 1]
```diff
@@ -294,6 +296,12 @@ public boolean appliesTo(Row row)
             return operator.isSatisfiedBy(column.type, rowValue(row), value);
         }
 
+        @Override
+        public boolean isNull(Row row)
+        {
+            return column.type.isNull(rowValue(row));
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/cql3/conditions/ColumnCondition.java:302]: `isNull` has the following semantics

1) for non-meangingless emptyness `null` and `empty` are both `null`
2) for meangingless emptyness `null` and `empty` are both `null` 
3) for types that support empty, only `null` is `null`

The main difference between 1/2 is that in 1 `empty` is rejected during validation, so normally means you passed a tombstone here by mistake (historic bug on the read path in `RowFilter`).  2 allows `empty` during validation

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java [hunk 2]
```diff
@@ -289,54 +291,99 @@ public boolean applies(TxnData data)
...
+                            // Collections don't support NULL but meangingless null types are supported, so byte[0] is allowed!
+                            // This is NULL when touched, so need to still check each value
+                            if (exists)
+                            {
+                                CollectionType<?> type = (CollectionType<?>) reference.column().type.unwrap();
+                                switch (type.kind)
+                                {
+                                    case MAP:
+                                    {
+                                        exists = !type.nameComparator().isNull(cell.path().get(0));
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: im still not 100% sold on this, here is the test that hits this

```
map<int, int>

IF row.map[<empty>] IS NULL -- returns true as the key is "null".  The data has map[<empty>] -> 42
IF row.map[42] IS NULL -- returns true if the value is <empty>
```

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java [hunk 3]
```diff
@@ -532,7 +589,13 @@ else if (type.isUDT())
         @Override
         public boolean applies(TxnData data)
         {
-            return getBounds(data).appliesTo(reference.getRow(data));
+            Bound bounds = getBounds(data);
+            if (reference.column().type.unwrap().isNull(bounds.value))
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java:580]: semantic difference with CAS IF.  if input or value are null, applies is

```
reference.column().type.unwrap().isNull(bounds.value) == bounds.isNull(row)
```

File: src/java/org/apache/cassandra/service/accord/txn/TxnReference.java [hunk 4]
```diff
@@ -351,4 +407,100 @@ public long serializedSize(TxnReference reference, TableMetadatas tables, Versio
             return size;
         }
     };
+
+    private static class ClusteringColumnData extends AbstractCell<ByteBuffer>
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/txn/TxnReference.java:529]: this is kinda a hack... it was this or refactor the code more.  I could also special case clustering, but felt that this was best as it kept the logic simpler

File: src/java/org/apache/cassandra/service/accord/txn/TxnReferenceValue.java [hunk 5]
```diff
@@ -60,9 +62,10 @@ enum Kind
 
     public static class Constant extends TxnReferenceValue
     {
+        @Nullable
         private final ByteBuffer value;
 
-        public Constant(ByteBuffer value)
+        public Constant(@Nullable ByteBuffer value)
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/txn/TxnReferenceValue.java:67]: should add a test to show this, but the value can be null which leads to NPE later on, so marked this nullable and made it handled

File: src/java/org/apache/cassandra/service/accord/txn/TxnReferenceValue.java [hunk 6]
```diff
@@ -115,19 +118,24 @@ void collect(TableMetadatas.Collector collector)
             @Override
             public void serialize(Constant constant, TableMetadatas tables, DataOutputPlus out, Version version) throws IOException
             {
-                ByteBufferUtil.writeWithVIntLength(constant.value, out);
+                out.writeBoolean(constant.value != null);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/txn/TxnReferenceValue.java:120]: not backwards or mixed compatible

File: test/unit/org/apache/cassandra/db/partitions/SimplePartition.java [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+import org.apache.cassandra.db.transform.FilteredRows;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.TimeUUID;
+import org.apache.cassandra.utils.btree.BTree;
+import org.apache.cassandra.utils.btree.UpdateFunction;
+import org.apache.cassandra.utils.memory.Cloner;
+import org.apache.cassandra.utils.memory.HeapCloner;
+
+public class SimplePartition extends AbstractBTreePartition
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/db/partitions/SimplePartition.java:50]: creating the `RowIterator` i need for the test was a lot of boiler plate, so figured could make this class to make writing tests less verbose

File: test/unit/org/apache/cassandra/db/partitions/SimplePartition.java [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+        }
+
+        public SimplePartition build()
+        {
+            SimplePartition.this.add(builder.build());
+            return SimplePartition.this;
+        }
+    }
+
+    private class Updater implements UpdateFunction<Row, Row>, ColumnData.PostReconciliationFunction
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/db/partitions/SimplePartition.java]: honestly copied a ton from atomic btree partition, the issue i faced with reusing is having an allocator...  I could try to make that logic reusable but didn't want to touch a critical section of code so I forked

File: src/java/org/apache/cassandra/cql3/conditions/ColumnCondition.java [hunk 9]
```diff
@@ -454,6 +473,12 @@ public boolean appliesTo(Row row)
             return operator.isSatisfiedBy((MultiElementType<?>) column.type, columnData, value);
         }
 
+        @Override
+        public boolean isNull(Row row)
+        {
+            return row == null || row.getComplexColumnData(column) == null;
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/cql3/conditions/ColumnCondition.java:479]: this is all filtered data, so can this data be non-live?

File: src/java/org/apache/cassandra/cql3/terms/Lists.java [hunk 10]
```diff
@@ -300,7 +300,7 @@ public Setter(ColumnMetadata column, Term t)
         @Override
         public boolean requiresTimestamp()
         {
-            return true;
+            return column.type.isMultiCell();
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/cql3/terms/Lists.java:303]: bug found in fuzz testing, frozen lists were requiring timestamp which made it hit this block

```
public void add(Operation operation, TableMetadata tableMetadata)
    {
        if (isForTxn && (operation.requiresRead() || operation.requiresTimestamp()))
        {
            add(operation.column, ReferenceOperation.create(operation, tableMetadata));
            return;
        }
```

this is a immutable bb so can avoid converting to a reference (i think it fails as a reference but forgot)

File: src/java/org/apache/cassandra/service/accord/AccordResult.java [hunk 11]
```diff
@@ -142,6 +143,8 @@ else if (isTxnRequest && coordinationFailed instanceof TopologyMismatch)
             }
             else
             {
+                logger.error("Unexpected exception", fail);
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/AccordResult.java:146]: I was seeing that logs were clean but exceptions were happening, and found that this code block was hiding all these exceptions.  The error isn't a `CoordinationFailed` so its not expected, so we should be loud!  I found several unexpected exceptions by adding this in

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java [hunk 12]
```diff
@@ -289,54 +291,99 @@ public boolean applies(TxnData data)
...
+                        // This is frozen, so check if the Cell is a tombstone and that the element is present.
+                        Cell<?> cell = (Cell<?>) columnData;
+                        exists = exists(cell, reference.column().type);
+                        if (exists)
+                        {
+                            ByteBuffer element = reference.getFrozenCollectionElement(cell);
+                            exists = !reference.getFrozenCollectionElementType().isNull(element);
+                        }
+                    }
+                    else if (reference.isFieldSelection())
```
Comment (dcapwell) [inline src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: im not sure if this is allowed in CQL, but it was already here in the code so i made sure it worked

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java (hunk 3 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java:580]: I *think* this is OK, because bound cannot be an IS NULL or IS NOT NULL check.

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java (hunk 2 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: I don't think we have a particularly good principled behaviour we can implement here, but I think treating empty as NULL wherever it is safe to do so, and unsafe to treat it as a raw value, is perhaps as good as we can get.

Comment (maedhroz) [reply src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: I haven't thought about this in a while, but can we just reject the txn if we use a meaningless key? If we can't, then agree... `map[null]` on a map w/ `int` keys is nonsense and should evaluate to `true` on `IS NULL`

Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: I think the problem is when we interact with data that already contains it. Literals perhaps can and should be rejected. Perhaps for transactions we can declare that these values are also generally unsupported even if the data is already there 🤷‍♀️

File: test/unit/org/apache/cassandra/db/partitions/SimplePartition.java (hunk 8 above)
Comment (belliottsmith) [reply test/unit/org/apache/cassandra/db/partitions/SimplePartition.java]: What was the issue with having an allocator?

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java [hunk 13]
```diff
@@ -289,54 +291,99 @@ public boolean applies(TxnData data)
             FilteredPartition partition = reference.getPartition(data);
             boolean exists = partition != null && !partition.isEmpty();
 
-            Row row = null;
-            if (exists)
+            if (reference.column() != null
+                && !reference.column().isPartitionKey())
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: Funny that the static analysis tools complain about this leaving NPE possible, but then I change it to...

```
ColumnMetadata column = reference.column();
if (column != null && !column.isPartitionKey())
```
...and it's apparently totally fine.

File: src/java/org/apache/cassandra/service/accord/txn/TxnReference.java [hunk 14]
```diff
@@ -43,18 +52,22 @@
...
 import static org.apache.cassandra.service.accord.AccordSerializers.columnMetadataSerializer;
 
 public class TxnReference
 {
     private final int tuple;
+    @Nullable
     private final TableMetadata table;
+    @Nullable
     private final ColumnMetadata column;
+    @Nullable
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/accord/txn/TxnReference.java]: I guess these are all technically nullable (for tests or w/ CAS requests?), but annotating them as such creates some noise for static analysis tools on this class. I wonder how hard it would be to make `TxnReference` and interface and separate the behavior of the normal case from the CAS case....

File: src/java/org/apache/cassandra/service/accord/txn/TxnReferenceValue.java [hunk 15]
```diff
@@ -90,7 +93,7 @@ public int hashCode()
         @Override
         public String toString()
         {
-            return "Constant=" + ByteBufferUtil.bytesToHex(value);
+            return "Constant=" + value == null ? "null" : ByteBufferUtil.bytesToHex(value);
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/accord/txn/TxnReferenceValue.java:95]:
```suggestion
            return "Constant=" + ((value == null) ? "null" : ByteBufferUtil.bytesToHex(value));
```

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java [hunk 16]
```diff
@@ -532,7 +589,13 @@ else if (type.isUDT())
         @Override
         public boolean applies(TxnData data)
         {
-            return getBounds(data).appliesTo(reference.getRow(data));
+            Bound bounds = getBounds(data);
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java:579]: nit: OCD in me is wondering if we can just change the method to singular `getBound()`

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java [hunk 17]
```diff
@@ -460,6 +512,11 @@ public Value(TxnReference reference, Kind kind, ByteBuffer value, ProtocolVersio
             this.version = version;
         }
 
+        public static EnumSet<Kind> supported()
+        {
+            return EnumSet.copyOf(KINDS);
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java:506]: nit: I guess either mark this `@VisibleForTesting` or just use `Sets.immutableEnumSet()` for the field?

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java [hunk 18]
```diff
@@ -289,54 +291,99 @@ public boolean applies(TxnData data)
...
+                                        exists = !type.nameComparator().isNull(cell.path().get(0));
+                                    }
+                                    break;
+                                    case LIST:
+                                    {
+                                        exists = !type.valueComparator().isNull(cell.buffer());
+                                    }
+                                    break;
+                                    default:
+                                        throw new UnsupportedOperationException(type.kind.name());
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: Do you think we could just fold this into `CollectionType` a la `makeCollectionReceiver()`? That might avoid the whole default branch throwing situation.

**Review by maedhroz — APPROVED** [review]

File: test/unit/org/apache/cassandra/db/partitions/SimplePartition.java (hunk 8 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/db/partitions/SimplePartition.java]: I honestly don't recall what was the issue i faced... ill relook at it

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java (hunk 13 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: This was a pet pev of mine in kotlin (the same people who made those tool you reference)... between the first and second call to `column` it might become null... by storing to a variable, you remove the possibility.

This is an issue with the static analysis, it doesn't really understand the callers and really doesn't grasp that this is basically immutable.  

I am 100% fine changing though, less warnings the better

File: src/java/org/apache/cassandra/service/accord/txn/TxnReference.java (hunk 14 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/txn/TxnReference.java]: not a bad idea, solving at the type level is always good... we have 3 cases?

1) row ref
2) column ref
3) sub-column ref

these are null for `row ref` only

Ill take a look at how painful its to refactor

File: test/unit/org/apache/cassandra/db/partitions/SimplePartition.java (hunk 8 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/db/partitions/SimplePartition.java]: switched to `AtomicBTreePartition` and the test is passing at least

File: src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java (hunk 13 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/txn/TxnCondition.java]: pushed the fix.  The code is more readable with the fix than without it, so its still a pro

File: src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java [hunk 19]
```diff
@@ -511,6 +512,14 @@ public ResultMessage execute(QueryState state, QueryOptions options, Dispatcher.
         {
             checkTrue(returningSelect.select.getLimit(options) == DataLimits.NO_LIMIT, NO_PARTITION_IN_CLAUSE_WITH_LIMIT, "SELECT", returningSelect.select.source);
         }
+        if (returningReferences != null)
+        {
+            for (RowDataReference dataRef : returningReferences)
+            {
+                TxnReference ref = dataRef.toTxnReference(options);
+                checkTrue(ref.kind == TxnReference.Kind.COLUMN, SELECT_WITH_REFERENCES_MUST_BE_A_COLUMN + ref.kind);
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java]: nit: Worth making the error `public` and throwing a quick test in `TransactionStatementTest` to hit this?

File: src/java/org/apache/cassandra/service/accord/txn/TxnReference.java [hunk 20]
```diff
@@ -163,192 +140,485 @@ public Row getRow(TxnData data)
...
+            return tuple.size() > field ? tuple.get(field) : null;
+        }
 
-        if (columnData == null)
-            return null;
+        public AbstractType<?> getFieldSelectionType()
+        {
+            assert isFieldSelection() : "No field selection type exists";
+            UserType userType = (UserType) column.type;
+            int field = ByteBufferUtil.getUnsignedShort(path.get(0), 0);
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/accord/txn/TxnReference.java:417]: Even though `path` is `@Nullable`, I think here and in a couple places directly above, all callers are still in situations where `path` should be non-null. Not sure if we want to make some assertions here, but things would blow up when misused either way.

**Review by maedhroz — APPROVED** [review]

File: src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java (hunk 19 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java]: turns out this is caught during `prepare`

```
if (returning != null)
            {
                // TODO: Eliminate/modify this check if we allow full tuple selections.
                returningReferences = returning.stream().peek(raw -> checkTrue(raw.column() != null, SELECT_REFS_NEED_COLUMN_MESSAGE))
                                                        .map(RowDataReference.Raw::prepareAsReceiver)
                                                        .collect(Collectors.toList());
            }
```

Comment (dcapwell) [reply src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java]: and tested here `org.apache.cassandra.cql3.statements.TransactionStatementTest#shouldRejectEntireTupleSelect`

Comment (dcapwell) [reply src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java]: removed this assert as its already handled

File: src/java/org/apache/cassandra/service/accord/txn/TxnReference.java (hunk 20 above)
Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/txn/TxnReference.java:417]: > Not sure if we want to make some assertions here, but things would blow up when misused either way.

We do

```
assert isFieldSelection() : "No field selection type exists";
```

that boils down to

```
public boolean selectsPath()
        {
            return path != null;
        }
```

In this case its the fact the type checker is dumb.  `path` has been proven to not be null assuming `assert` isn't disabled

Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/txn/TxnReference.java:417]: actually for me `path` isn't what it complains about, its `get(0)`

Comment (dcapwell) [reply src/java/org/apache/cassandra/service/accord/txn/TxnReference.java:417]: its still wrong as we are either `empty` or `single`; and `single` doesn't allow `null`

```
public static CellPath create(ByteBuffer value)
    {
        assert value != null;
        return new SingleItemCellPath(value);
    }
```

**Comment by dcapwell** [comment]
this was merged
